lib/telemetry_metrics_mnesia/db.ex

```diff
@@ -43,36 +67,112 @@ defmodule TelemetryMetricsMnesia.Db do
     end
   end
 
-  def fetch(%_{tags: tags, keep: keep} = metric) do
-    default =
-      case tags do
-        [] ->
-          0
+  def fetch(metric) do
+    @telemetry_metrics_table
+    |> Mnesia.dirty_read(metric)
+    |> case do
+      [] ->
+        default(metric)
 
-        _ ->
-          %{}
-      end
+      [telemetry_metric] ->
+        telemetry_metrics(telemetry_metric, :value)
+    end
+  end
+
+  def apply_metric_type(data, %mod{} = metric) when mod in [Distribution, Summary] do
+    reducer =
+      metric
+      |> events_reducer_fun()
+      |> check_granularity(metric)
 
+    data
+    |> Enum.reduce_while(%{}, reducer)
+    |> Map.new(stat_fun(mod))
+    |> case do
+      %{%{} => data} ->
+        data
+
+      data ->
+        data
+    end
+  end
+
+  def apply_metric_type(data, metric) do
     metric
-    |> fetch_events()
-    |> Enum.filter(&keep?(&1, keep))
-    |> reduce_events(metric, events_reducer_fun(metric), default)
+    |> events_reducer_fun()
+    |> check_granularity(metric)
+    |> then(&Enum.reduce_while(data, default(metric), &1))
+  end
+
+  defp check_granularity(reducer, %_{reporter_options: []}), do: reducer
+
+  defp check_granularity(reducer, metric) do
+    case granularity_min_timestamp(metric) do
+      nil ->
+        reducer
+
+      timestamp ->
+        fn
+          {ts, _v, _tag_values}, acc when ts <= timestamp ->
+            {:halt, acc}
+
+          v, acc ->
+            reducer.(v, acc)
+        end
+    end
+  end
+
+  defp granularity_min_timestamp(%_{reporter_options: opts}) do
+    case Keyword.get(opts, :granularity) do
+      [{unit, amount}] ->
```

Comment:
Why is granularity a single element list? Should it be `{unit, amount}` instead? Is it possible to have >1 element?

Reply:
ah, I didn't realize, makes sense!